### PR TITLE
New complaint list page

### PR DESCRIPTION
**Pull request by rpadovani**
### Checklist
- [x] I run `yarn run webpack:build:main`: the project builds without errors.
- [x] I run `yarn lint`: the project builds without code style warnings.
- [x] ~I updated the documentation and models.~
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] ~I added (end-to-end) test cases for the new functionality.~

### Motivation and Context

This pull request creates a new component for listing complaints, that it can be used by tutors to see complaints about themselves, or by instructors to see complaints in a course / exercise.

This implements [ARTEMIS-348](https://jirabruegge.in.tum.de/browse/ARTEMIS-348) and [ARTEMIS-349](https://jirabruegge.in.tum.de/browse/ARTEMIS-349)

### Description

I introduced a new complaint service to manage all the different logic about complaints, and I have also rewritten the complaint resource to use only the complaint service and not the complaint repository directly to guarantee proper encapsulation and separation of concerns.

I have also noticed that we weren't actually calculating the number of complaints about a single tutor in the course dashboard, so I fixed that as well.

### Steps for Testing

As a tutor, you can access to complaints about you in two ways:
- From a tutor course dashboard, you can click on "Complaints about your assessments" and you will see complaints about you for that course
- From a tutor exercise dashboard, you can click on "Complaints about your assessments" and you will see complaints about you for that exercise

As an instructor, there are different ways:
-  From the instructor course dashboard, you click on "complaints" to see all the complaints in the course
-  From the instructor course dashboard, you go to the tutor leaderboard and click on the number of complaints near to a tutor, to see all the complaints about that tutor in the course
-  From the instructor exercise dashboard, you click on "complaints" to see all the complaints in the exercise
-  From the instructor exercise dashboard, you go to the tutor leaderboard and click on the number of complaints near to a tutor, to see all the complaints about that tutor in the exercise

### Screenshots

image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java [hunk 1]
```diff
@@ -392,4 +392,13 @@ public User updateUserNotificationReadDate() {
         return loggedInUser;
     }
 
+    @Transactional(readOnly = true)
+    public User getUserById(Long tutorId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/UserService.java]: Can you not use getUserWithAuthorities?
Definitely make sure tutorId --> userId

**Comment by krusche** [comment]
Looks good to me in general. Some improvement ideas:
1) Please make the whole Complaints String (next to the number) clickable as link and not only the complaint number itself
2) Make sure to specify whose login and name you are showing! Is this the student or the reviewer? Can we show both for instructors?

**other important notes:** 
3) Do you make sure on the server side that tutors can only see the complaints about their own assessments? 
4) Do you make sure on the server side that tutors can not see the student names? I think they should be able to see the original reviewer (tutor) so that they could communicate with this tutor in case something is unclear

Thanks

**Comment by krusche** [comment]
@SiggZ you have worked with complaints before, please review the PR and test if everything still works and we don't have any regressions. Thanks

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java (hunk 1 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/service/UserService.java]: You are right, I overthought this one. I removed the method and simplified the code where it was used

**Comment by rpadovani** [comment]
> Do you make sure on the server side that tutors can only see the complaints about their own assessments?

Yes, in the new resources I created for this task. Of course tutors can still go submission by submission to check them

**Comment by rpadovani** [comment]
> Please make the whole Complaints String (next to the number) clickable as link and not only the complaint number itself

Done

> Make sure to specify whose login and name you are showing! Is this the student or the reviewer? Can we show both for instructors?

Fixed: specify it is the reviewer, and add the students for the instructors (checks are both on server and on client)

> 4) Do you make sure on the server side that tutors can not see the student names? I think they should be able to see the original reviewer (tutor) so that they could communicate with this tutor in case something is unclear

Yes, students are filtered out for tutors, but not for instructors. They see the original reviewer on the top of the page when they open the submission.

Please notice I haven't tested this new version on the test server yet

**Comment by rpadovani** [comment]
Tested also this new version on the test server, it seems to work as expected

**Comment by krusche** [comment]
I already merged it, please test it before the next bugfix release again as part of a develop release

**Pull request merged**